File: src/main/python/graph/course_schedule.py
from typing import *


class CourseSchedule:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        def construct():
            adj = {k:[] for k in range(numCourses)}
            for required, course in prerequisites: adj[course].append(required)
            return adj

        graph = construct()
        visited = [0] * numCourses
        def has_cycle(index):
            if visited[index] == 1: return False
            if visited[index] == -1: return True
            visited[index] = - 1
            for n in graph[index]:
                if has_cycle(n): return True
            visited[index] = 1

        for i in range(numCourses):
            if has_cycle(i):
                return False
        return True


if __name__ == '__main__':
    s = CourseSchedule()
    r = s.canFinish(20, [[0, 10], [3, 18], [5, 5], [6, 11], [11, 14], [13, 1], [15, 1], [17, 4]])
    print(r)

File: src/main/java/org/parthvnp/Graph/CourseSchedule.java
package org.parthvnp.Graph;

import java.util.*;

public class CourseSchedule {
  public HashMap<Integer, ArrayList<Integer>> graph = new HashMap<>();
  public int[] visited;

  public boolean canFinish(int numCourses, int[][] prerequisites) {
    visited = new int[numCourses];
    construct(prerequisites);
    for (int i = 0; i < numCourses; i++) if (hasCycle(i)) return false;
    return true;
  }

  public void construct(int[][] prereqs) {
    for (int[] i : prereqs) {
      graph.computeIfAbsent(i[1], k -> new ArrayList<>());
      graph.computeIfPresent(
          i[1],
          (k, v) -> {
            v.add(i[0]);
            return v;
          });
    }
  }

  public boolean hasCycle(int index) {
    if (visited[index] == 1) return false;
    if (visited[index] == -1) return true;
    visited[index] = -1;
    for (int i : graph.getOrDefault(index, new ArrayList<>())) {
      if (hasCycle(i)) return true;
    }
    visited[index] = 1;
    return false;
  }

  public static void main(String[] args) {
    var s = new CourseSchedule();
    var r = s.canFinish(2, new int[][]{new int[]{1,0}});
    System.out.println(s.graph);
    System.out.println(r);
  }
}
